### ISSUE TITLE
Session timeout

### ISSUE THREAD

**Issue by xchs** [contao/core-bundle#1428]
I have set the session timeout to 360000 seconds so that the (back end) session should be still valid on the next day:

`system/config/localconfig.php`:
```php
$GLOBALS['TL_CONFIG']['sessionTimeout'] = 360000;
```

However, if I click on a menu item in the back end the next day, I am always logged out and have to log in again.

[Contao 4.4.16 / Chrome, Firefox]

**Comment by aschempp** [comment]
Looks like your IP address changes after a day, which is often the case with DSL providers.

**Comment by xchs** [comment]
I wil check that.

**Comment by xchs** [comment]
No, I'm afraid it's not. The IP address has been the same for over a month.

**Comment by leofeyer** [comment]
Did you also adjust the timeout of the PHP session cookie in your php.ini? Why is this desirable in the first place?

**Comment by xchs** [comment]
> Did you also adjust the timeout of the PHP session cookie in your php.ini?

I have not changed anything in the PHP configuration. This is how the current configuration looks like:

![php_session](https://user-images.githubusercontent.com/1218766/37458856-99766b40-2846-11e8-92d5-74081aec06ec.jpg)

The `session.cookie_lifetime` is set to `0`, which means the session should only be valid until the next browser restart. However, if I have opened the back end in the browser, close the browser and then open it again, I can easily continue working in the back end without having to log in again.

**Comment by fritzmg** [comment]
@xchs were you able to solve the issue?

**Comment by xchs** [comment]
No. However, for the time being I have closed the ticket for two reasons:

- > If you feel like you have to adjust the PHP session timeout, your php.ini would be the correct place to do so.
- I would like to investigate further whether there is a difference between Contao 3 and Contao 4 in this regard.

**Comment by xchs** [comment]
> I would like to investigate further whether there is a difference between Contao 3 and Contao 4 in this regard.

That's what I did now. There actually seems to be differences here. While the session of a Contao 3.5 installation was still active the next day, the session of a Contao 4.4 installation had apparently expired. Same settings, same host.

**Comment by fritzmg** [comment]
So this needs to be reopened? ;)

**Comment by leofeyer** [comment]
Only if you know why this is and how to fix it.

**Comment by asaage** [comment]
I guess it needs to be reopened 🤣 
1. it has been closed without a fix
2. i wouldn't take into account the question why this is desirable in the first place - the session expires after exactly 1h although the setting states it shoud do something different.

If this is happening for every environment and it is not fixable you could also eaasily ommit this setting.

The fact that a ```session.cookie_lifetime = 0``` does not terminate a session on browser restart is confusing but i'm used to it by other sites.

The only thing with a lifetime of 1h i could find is the ```csrf_https-contao_csrf_token```-cookie but if i delete it i am still logged in.

**Comment by aschempp** [comment]
The backend login time in Contao 4.4 is not related to the session timeout, because we're storing our own cookie. (It completely changed in Contao 4.5 though!)

**Comment by asaage** [comment]
I'm using 4.5.8
Is the ```System>Settings>Timeout values>Session timeout``` value for the FrontEnd only? (because the other three field below this legend are BE-related)
How can you tweak this for the BE?

**Comment by asaage** [comment]
any news on this one?

**Comment by Aybee** [comment]
I also would like to know if this sessionTimeout regards to the BE, FE or both.

**Comment by joehi** [comment]
I have the problem that after round about half an hour of idle time I always have to newly login to the BE. I increased the session timeout in the settings from 3600 to 36000 seconds but that did not help at all. As long as I am active in the BE there are not problems. But as soon as I have to interrupt I run into this problem.
This occurs with Contao 4.4.26 under php7.0 and after upgrading to php7.2 I have the same effect with both  Contao 4.4.26 and Contao 4.6.5 respectively.
Yesterderday did some tests with intentionally idling and measuring the time. With up to 30 minutes idle time I had not problem and could continue without having to re-login to the BE. With an idle time of 35 minutes the session expired and I had to re-login to the BE. Between 30 and 35 minutes idle time I had both results.

Disabling the ip-check did not solve this problem, nor did it help to disable the query token. My IP address does not change.

**Comment by aschempp** [comment]
In Contao before 4.5, the backend session timeout is configured in the backend settings.

**Comment by joehi** [comment]
I had increased the session timeout in the backend from 3600 to 36000 seconds but that does not have any influence. I do not even reach the 60 minutes at all.

**Comment by tabcontao** [comment]
Researching the web it is clear, that Symfony does offer the functionality to invalidate/delete the session after a settable time of inactivity. This is independent of the global session timeout! I did not find out yet if this is active by default or if it is activated by Contao (4.6.5 in my case) and which default idle time is used if nothing is set by the application. Anyway, in my case it seems to be 30 minutes. And it seems not to be related to the cookie lifetime. Doing something in the backend after 29 minutes avoids the auto-logout. Did this 3 times in a row and still no logout (BE settings are set to 36000 seconds for the tests).

**Comment by aschempp** [comment]
Please be aware that there are two diffenrent implementations to consider!

Contao 4.4 (every version before 4.5) has a custom **user session** implementation, which is not related to the **PHP session** at all! The session is persisted in the `BE_USER_AUTO` cookie, which has a lifetime according to the backend settings. 

In Contao 4.5 and later, we switched the user authentication to use the full regular Symfony implementation. There is no `BE_USER_AUTH` cookie anymore, the basic lifetime is bound to the **PHP session timeout**. This can be configured in your PHP config, and the session is persisted in the `PHPSESSION` cookie (by default).
Additionally, we implemented a custom timeout handling, because some browsers (like Chrome) ignore a cookie timeout and keep the session cookie for longer that they should (probably to improve user experience by keeping you logged in).

Whatever you read about the Symfony implementation does not apply to Contao 4.4. Whatever experience you have with the previous implementation (e.g. from Contao 3.5) does not apply to Contao 4.5/4.6.

**Comment by tabcontao** [comment]
Thanks for the clarification. In a way, this provides an explanation for my results with Contao 4.6. I guess the PHP session timeout is controlled by the PHP configuration options

session.gc_maxlifetime
session.gc_probability
session.gc_divisor

Am I right here or are there other relevant options?
Will sessions started by other users running scripts on the same apache instance with the same PHP version also count as new sessions?

But, this still does not explain the experience of Joehi with his Contao 4.4, which is virtually the same as with Contao 4.6 - logout after about 30-35 minutes of inactivity.

Anyway, I'll give it a try on another hosting, where I am in control of all that. The plan for testing would be to lower the value of gc_maxlifetime to e.g. 60, setting gc.probability and gc_divisor to 1. This should reduce the session lifetime significantly with Contao 4.5 and 4.6. But with Contao 3.5 or 4.4 nothing should change?

**Comment by xchs** [comment]
> In Contao 4.5 and later, we switched the user authentication to use the full regular Symfony implementation. There is no `BE_USER_AUTH` cookie anymore, the basic lifetime is bound to the **PHP session timeout**.


But that would mean, however, that from Contao >= 4.5 on the corresponding option in the system settings has no effect anymore and should actually be removed, right?

>  This can be configured in your PHP config, and the session is persisted in the `PHPSESSION` cookie (by default).

However, many will probably not be able to change the PHP settings on their own.

**Comment by aschempp** [comment]
> But that would mean, however, that from Contao >= 4.5 on the corresponding setting in the system settings has no effect anymore and should actually be removed, right?

> However, many will probably not be able to change the PHP settings on their own.

No, and not necessary, because

>> Additionally, we implemented a custom timeout handling, because some browsers (like Chrome) ignore a cookie timeout and keep the session cookie for longer that they should (probably to improve user experience by keeping you logged in).

**Comment by tabcontao** [comment]
> Additionally, we implemented a custom timeout handling, because some browsers (like Chrome) ignore a cookie timeout and keep the session cookie for longer that they should (probably to improve user experience by keeping you logged in).
Would this custom timeout handling explain the auto-logout after about 30 minutes? If yes, can the timeout value be changed by a configuration parameter and why would this auto-logout (according to Joehi) also happen with Contao 4.4?

**Comment by joehi** [comment]
It seems that I found the reason for the timeout problems. It is the following entry in the php.ini:
`session.gc_maxlifetime = 1440`
After changing the default 1440 (seconds) to 7200 I did not get the idle timout after 30 minutes. Now I could stay logged in even with an idle time of 2 hours. After 3 hours of idle time my sessions were expired, but that was expected with a setting of 7200 seconds = 2 hours. And now the values in the Contao Backend are limiting the session if they are set to for example 2 hours.

This is valid both for Contao 4.4.26 and Contao 4.6 as well.

It would be helpful if there could be a hint in the Contao Backend in the help text of the value for the session timeout. It could mention that this setting in the php.ini could be the limiting factor.
